### ISSUE TITLE
Energy derivatives wrt nuclear coordinates with order > 2 disagrees with finite differences

### ISSUE THREAD

**Issue by chertianser** [fishjojo/pyscfad#61]
I'm attempting to benchmark the accuracy of auto-differentiated higher order derivatives with respect to energy. The chunk of code below will show that:

1.  Finite differences of energy `grad_finite_derives` agrees with the gradient from autodiff  `grad_ad`:
    ```
    np.isclose(grad_finite_derives, grad_ad, atol=1e-5).all()
    >>> True
    ```
2.  Finite differences of gradients `hess_finite_derives` agrees with the Hessian from autodiff `hess_ad`:
    ```
    np.isclose(hess_finite_derives, hess_ad, atol=1e-5).all()
    >>> True
    ```
3.  Finite differences of Hessians `three_finite_derives` **_disagree_** with the 3rd order derivative from autodiff `three_ad`.

    ```
    np.isclose(three_finite_derives, three_ad, atol=1e-5).all()
    >>> False
    ```
This behavior is not sensitive to the size of the finite difference step (tested 1e-3, 1e-5 and 1e-7).

`three_finite_derives` will show that entries in the tensor are symmetrical and opposite:

```
three_finite_derives
>>> array([[[[[[ 0.        ,  0.        , -0.01280354],
           [ 0.        ,  0.        ,  0.01280354]],

          [[ 0.        ,  0.        ,  0.        ],
           [ 0.        ,  0.        ,  0.        ]],

          [[-0.01280354,  0.        ,  0.        ],
           [ 0.01280354,  0.        ,  0.        ]]],
          ...
```

while the auto-diffed `three_ad` shows that some values agree with the finite deriviatives, but others will deviate:

```
three_ad
>>> array([[[[[[[ 0.        ,  0.        , -0.01280354],
            [ 0.        ,  0.        ,  0.01280354]],

           [[ 0.        ,  0.        ,  0.        ],
            [ 0.        ,  0.        ,  0.        ]],

           [[ 0.30266442,  0.        ,  0.        ],
            [ 0.01280354,  0.        ,  0.        ]]],
           ...
```

This is the minimum viable example which will show this behavior:

```
import numpy as np
import jax
from pyscfad import gto, scf

def build_mol(atom):
    mol = gto.Mole()
    mol.atom = atom
    mol.unit = 'Angstrom'
    mol.basis = 'sto-3g'
    mol.build(trace_coords=True, trace_exp=True, trace_ctr_coeff=True)
    return mol

def ehf(mol):
    mf = scf.RHF(mol)
    e = mf.kernel()
    return e

def generate_finite_differences(atom_string, d):
    # Parse the atom string into a list of atoms and their coordinates
    atoms = [atom.strip().split() for atom in atom_string.split(';')]
    
    # Generate finite differences
    results = []
    for i, atom in enumerate(atoms):
        for j in range(1, 4):  # For each coordinate (x, y, z)
            for sign in [1, -1]:
                new_atoms = [a[:] for a in atoms]  # Deep copy of atoms
                new_atoms[i][j] = f"{float(atom[j]) + sign * d:.7f}"
                
                # Format the result string
                result = '; '.join([f"{a[0]} {a[1]} {a[2]} {a[3]}" for a in new_atoms])
                results.append(result)
    
    return results

def derivative(plus, minus, d):
    return (plus - minus) / (2 * d) * 0.529177

def get_finite_derivatives(finite_diff, d):
    all_deriv = []
    for coord in finite_diff:
        deriv_list = []
        for i in range(0, coord.shape[0], 2):
            deriv = derivative(coord[i], coord[i+1], d)
            deriv_list.append(deriv)
        all_deriv.append(deriv_list)
        
    return np.array(all_deriv)

grad_list = []
hess_list = []
three_list = []

grad_finite_diff = []
hess_finite_diff = []
three_finite_diff = []

d = 1e-5

atom = f"H 0 0 0; H 0 0 1.0"

mol = build_mol(atom)
_ , grad = jax.value_and_grad(ehf)(mol)
hess = jax.jacfwd(jax.jacrev(ehf))(mol)
three = jax.jacfwd(jax.jacrev(jax.jacfwd(ehf)))(mol)
grad_list.append(grad)
hess_list.append(hess)
three_list.append(three)

finite_differences = generate_finite_differences(atom, d)
grad_diff_list = []
hess_diff_list = []
three_diff_list = []
for diff in finite_differences:
    fd_mol = build_mol(diff)
    grad_diff_list.append(ehf(fd_mol))
    hess_diff_list.append(jax.jacfwd(ehf)(fd_mol))
    three_diff_list.append(jax.jacfwd(jax.jacfwd(ehf))(fd_mol))
grad_finite_diff.append(grad_diff_list)
hess_finite_diff.append(hess_diff_list)
three_finite_diff.append(three_diff_list)

grad_finite = np.array([[float(x) for x in y] for y in grad_finite_diff])
grad_ad = np.array([x.coords for x in grad_list])

hess_finite = np.array([[x.coords for x in y] for y in hess_finite_diff])
hess_ad = np.array([x.coords.coords for x in hess_list])

three_finite = np.array([[x.coords.coords for x in y] for y in three_finite_diff])
three_ad = np.array([x.coords.coords.coords for x in three_list])

grad_finite_derives = get_finite_derivatives(grad_finite, d)
grad_finite_derives = grad_finite_derives.reshape(2,3)
hess_finite_derives = get_finite_derivatives(hess_finite, d)
hess_finite_derives = hess_finite_derives.reshape(2,3,2,3)
three_finite_derives = get_finite_derivatives(three_finite, d)
three_finite_derives = three_finite_derives.reshape(2,3,2,3,2,3)

print(np.isclose(grad_finite_derives, grad_ad, atol=1e-5).all())
print(np.isclose(hess_finite_derives, hess_ad, atol=1e-5).all())
print(np.isclose(three_finite_derives, three_ad, atol=1e-5).all())
```
